ndb/src/google/cloud/ndb/model.py

```diff
@@ -336,6 +339,124 @@ def __hash__(self):
+    This is what is meant by composable (or stackable) APIs.
+
+    The API supports "stacking" classes with ever more sophisticated
+    user <-> base conversions: the user -> base conversion goes from more
+    sophisticated to less sophisticated, while the base -> user conversion goes
+    from less sophisticated to more sophisticated. For example, see the
+    relationship between :class:`BlobProperty`, :class:`TextProperty` and
+    :class:`StringProperty`.
+
+    In addition to ``_to_base_type()`` and ``_from_base_type()``, the
+    ``_validate()`` method is also a composable API.
+
+    The validation API distinguishes between "lax" and "strict" user values.
+    The set of lax values is a superset of the set of strict values. The
+    ``_validate()`` method takes a lax value and if necessary converts it to
+    a strict value. This means that when setting the property value, lax values
+    are accepted, while when getting the property value, only strict values
+    will be returned. If no conversion is needed, ``_validate()`` may return
+    :data`None`. If the argument is outside the set of accepted lax values,
+    ``_validate()`` should raise an exception, preferably :exc:`TypeError` or
+    :exc:`.BadValueError`.
+
+    Example/boilerplate:
+
+    .. code-block:: python
+
+        def _validate(self, value):
+            # Lax user value to strict user value.
+            if not isinstance(value, <top type>):
+                raise TypeError(...)  # Or datastore_errors.BadValueError(...).
+
+        def _to_base_type(self, value):
+            # (Strict) user value to base value.
+            if isinstance(value, SomeUserType):
+                return SomeBaseType(value)
+
+        def _from_base_type(self, value):
+            # Base value to (strict) user value.'
+            if not isinstance(value, SomeBaseType):
+                return SomeUserType(value)
+
+    Things that ``_validate()``, ``_to_base_type()`` and
+    ``_from_base_type()`` do **not** need to handle:
+
+    * :data:`None`: They will not be called with :data:`None` (and if they
+      return :data:`None`, this means that the value does not need conversion).
+    * Repeated values: The infrastructure (:meth:`_get_user_value` and
+      :meth:`_get_base_value`) takes care of calling ``_from_base_type()`` or
+      ``_to_base_type()`` for each list item in a repeated value.
+    * Wrapping values in :class:`_BaseValue`: The wrapping and unwrapping is
+      taken care of by the infrastructure that calls the composable APIs.
+    * Comparisons: The comparison operations call ``_to_base_type()`` on
+      their operand.
+    * Distinguishing between user and base values: the infrastructure
+      guarantees that ``_from_base_type()`` will be called with an
+      (unwrapped) base value, and that ``_to_base_type()`` will be called
+      with a user value.
+    * Returning the original value: if any of these return :data:`None`, the
```

Comment:
I'm uncomfortable with the use of `None` as a semaphore for "do nothing." However remote, it seems there could be a possibility of a validator intending to convert a value to `None`. In general, in these kinds of scenarios, I prefer an explicit, unambiguous semaphore, usually passed into the method that might use it:

def _from_base_type(self, value, _nope):
    # Base value to (strict) user value.'
    if not isinstance(value, SomeBaseType):
        return SomeUserType(value)
    return _nope

... but with a better name than `_nope`.  I suspect we're constrained by backwards compatibility considerations on this point, though, so this is probably more for posterity than anything else.

Reply:
Yes, we are constrained here. The documented behavior is that a return value of `None` in the MRO chain (of `_(to|from)_base_type()`) just means that transformation didn't make any changes.

The composition / stacking is done via `_call_(to|from)_base_type()`.